src/Pull.cpp

```diff
@@ -0,0 +1,200 @@
+                 const bool shouldConnect) {
+   mProtocolHandler = std::move(std::unique_ptr<NanoProtocol>(
+                                   new NanoProtocol(location)));
+   mSocket = nn_socket(AF_SP, NN_PULL);
+   if (mSocket < 0) {
+      throw std::runtime_error("could not open transport socket" +
+                               mProtocolHandler->GetLocation() +
+                               " because of error: " +
+                               std::string(nn_strerror(errno)));
+   }
+   nn_setsockopt(mSocket,
+                 NN_SOL_SOCKET,
+                 NN_RCVTIMEO,
+                 &timeoutInMs,
+                 sizeof (timeoutInMs));
+   int connectResponse = {0};
+   std::string connectionType("");
+   if (shouldConnect){
+      connectResponse = nn_connect(mSocket,
+                                   mProtocolHandler->GetLocation().c_str());
+      connectionType = "nn_connect";
+   } else{
+      connectResponse = nn_bind(mSocket,
+                                mProtocolHandler->GetLocation().c_str());
+      connectionType = "nn_bind";
+   }
+   if (connectResponse < 0) {
+      throw std::runtime_error("could not connect to endpoint: " +
+                               mProtocolHandler->GetLocation() +
+                               " because of error: " +
+                               std::string(nn_strerror(errno)));
+   }
+   LOG(INFO) << connectionType
+              << " socket at: " << mProtocolHandler->GetLocation()
+              << " with timeout: " << timeoutInMs;
+}
+/**
+ * Construct a non-blocking nanomsg pull (timeout)
+ */
+Pull::Pull(const std::string& location,
+           const int timeoutInMs,
+           const int bufferSize,
+           const bool shouldConnect) {
+   mProtocolHandler = std::move(std::unique_ptr<NanoProtocol>(
+                                   new NanoProtocol(location)));
+   mSocket = nn_socket(AF_SP, NN_PULL);
+   if (mSocket < 0) {
+      throw std::runtime_error("could not open transport socket" +
+                               mProtocolHandler->GetLocation() +
+                               " because of error: " +
+                               std::string(nn_strerror(errno)));
+   }
+   if (timeoutInMs > 0) {
+      nn_setsockopt(mSocket,
+                    NN_SOL_SOCKET,
+                    NN_RCVTIMEO,
+                    &timeoutInMs,
+                    sizeof(timeoutInMs));
+   } else {
+      LOG(INFO) << " zero or negative timeout requested, blocking socket";
+   }
+   nn_setsockopt(mSocket,
+                 NN_SOL_SOCKET,
+                 NN_RCVBUF,
+                 &bufferSize,
+                 sizeof(bufferSize));
+   int connectResponse = {0};
+   std::string connectionType("");
+   if (shouldConnect){
+      connectResponse = nn_connect(mSocket,
+                                   mProtocolHandler->GetLocation().c_str());
+      connectionType = "nn_connect";
+   } else{
+      connectResponse = nn_bind(mSocket,
```

Comment:
See comment below. I would much rather have logic extracted out of the constructors using factory methods.